sonic_platform_base/sonic_xcvr/api/public/cmis.py

```diff
@@ -283,3 +590,970 @@ def get_lpmode_support(self):
+        if lplonly_flag:
+            BLOCK_SIZE = 116
+        else:
+            BLOCK_SIZE = maxblocksize
+        address = 0
+        remaining = imagesize - startLPLsize
+        logger.info("\nTotal size: {} start bytes: {} remaining: {}".format(imagesize, startLPLsize, remaining))
+        while remaining > 0:
+            if remaining < BLOCK_SIZE:
+                count = remaining
+            else:
+                count = BLOCK_SIZE
+            data = f.read(count)
+            progress = (imagesize - remaining) * 100.0 / imagesize
+            if lplonly_flag:
+                fw_download_status = self.cdb.block_write_lpl(address, data)
+            else:
+                fw_download_status = self.cdb.block_write_epl(address, data, autopaging_flag, writelength)
+            if fw_download_status != 1:
+                logger.info('CDB download failed. CDB Status: %d' %fw_download_status)
+                exit(1)
+            elapsedtime = time.time()-starttime
+            logger.info('Address: {:#08x}; Count: {}; Progress: {:.2f}%; Time: {:.2f}s'.format(address, count, progress, elapsedtime))
+            address += count
+            remaining -= count
+        elapsedtime = time.time()-starttime
+        logger.info('Total module FW download time: %.2f s' %elapsedtime)
+
+        time.sleep(2)
+        # complete FW download (CMD 0107h)
+        fw_complete_status = self.cdb.validate_fw_image()
+        if fw_complete_status == 1:
+            logger.info('Module FW download complete: Success')
+        else:
+            logger.info('Module FW download complete: Fail')
+        elapsedtime = time.time()-elapsedtime-starttime
+        logger.info('Complete module FW download time: %.2f s\n' %elapsedtime)
+
+    def module_fw_upgrade(self, imagepath):
+        """
+        This function performs firmware upgrade.
+        1.  show FW version in the beginning
+        2.  check module advertised FW download capability
+        3.  configure download
+        4.  show download progress
+        5.  configure run downloaded firmware
+        6.  configure commit downloaded firmware
+        7.  show FW version in the end
+        """
+        self.get_module_fw_info()
+        startLPLsize, maxblocksize, lplonly_flag, autopaging_flag, writelength = self.get_module_fw_upgrade_feature()
+        self.module_fw_download(startLPLsize, maxblocksize, lplonly_flag, autopaging_flag, writelength, imagepath)
+        self.module_fw_run(mode = 0x01)
+        time.sleep(60)
+        self.module_fw_commit()
+        self.get_module_fw_info()
+
+    def module_fw_switch(self):
+        """
+        This function switch the active/inactive module firmware in the current module memory
+        """
+        _, _, _, ImageAValid, _, _, _, ImageBValid = self.get_module_fw_info()
+        if ImageAValid == 0 and ImageBValid == 0:
+            self.module_fw_run(mode = 0x01)
+            time.sleep(60)
```

Comment:
is this sleep from spec? where in spec?

Reply:
This is not in the spec. See description from CMIS. 60 seconds wait time is empirical for run command to finish. 

"Executing the Run Image command may potentially be hitless, non-disruptive or minimally disruptive to high6 speed traffic.
Behavior of the module and its control loop transients during resets are vendor and technology dependent."